R/layers.R

```diff
@@ -237,12 +241,120 @@ addMarkers = function(
   options = markerOptions(),
   data = getMapData(map)
 ) {
-  options$icon = icon
+  options$icon = L.icon(evalFormula(icon, data))
+  options$iconGroup = attr(options$icon, 'iconGroup', exact = TRUE)
   pts = derivePoints(data, lng, lat, missing(lng), missing(lat), "addMarkers")
   invokeMethod(map, data, 'addMarkers', pts$lat, pts$lng, layerId, options, popup) %>%
     expandLimits(pts$lat, pts$lng)
 }
 
+L.icon = function(options, embed = TRUE) {
+  if (!is.list(options)) return(options)
+  if (is.null(names(options))) {
+    if (length(options) == 0) return()
+    # in theory we need to check all sub-lists and all must be named
+    if (is.null(names(options[[1]])))
+      stop('The data for an individual icon must be a named list')
+    # it is already a list of icons, so no further base64 encoding
+    return(options)
+  }
+  image_uri = getFromNamespace('.b64EncodeFile', 'markdown')
+  for (i in c('iconUrl', 'iconRetinaUrl', 'shadowUrl', 'shadowRetinaUrl')) {
+    Url = options[[i]]
+    if (!embed || !is.character(Url) || !file.exists(Url)) next
+    options[[i]] = image_uri(Url)
+  }
+  options
+}
+
+#' Create a list of icons
+#'
+#' An icon can be represented as a list of the form \code{list(iconUrl,
+#' iconSize, ...)}. This function is vectorized over its arguments to create a
+#' list of icons. See \url{http://leafletjs.com/reference.html#icon} for the
+#' possible attributes of icons.
+#'
+#' Note some icon attributes are of length 2, such as \code{iconSize} and
+#' \code{iconAnchor}, and \code{iconList()} provides two separate arguments for
+#' these attributes, e.g. \code{iconSizeX = 20} and \code{iconSizeY = 40} means
+#' \code{iconSize = c(20, 40)} internally. Shorter argument values will be
```

Comment:
It was written for those who read the API reference http://leafletjs.com/reference.html#icon and I was concerned that they might be confused by the inconsistency of argument names, e.g. there is no `iconWidth`/`iconHeight` in the leafletjs documentation. Well, I guess it will be my problem if users really have to read the leafletjs documentation to understand these arguments, so I'll drop this sentence.